// browser-http/lib/Xhr.js
(function() {
  var EventEmitter, FakePromise, Helpers, Response, Xhr, escape,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Helpers = require('./Helpers');

  Response = require('./Response');

  FakePromise = require('./FakePromise');

  EventEmitter = require('events').EventEmitter;

  escape = require('escape-regexp');

  Xhr = (function(superClass) {
    extend(Xhr, superClass);

    Xhr.JSONP_METHOD_PREFIX = '__browser_http_jsonp_callback_';

    Xhr.COUNTER = 0;

    Xhr.prototype.xhr = null;

    Xhr.prototype.response = null;

    Xhr.prototype.url = null;

    Xhr.prototype.type = 'GET';

    Xhr.prototype.data = null;

    Xhr.prototype.jsonp = false;

    Xhr.prototype.jsonPrefix = null;

    function Xhr(url, type, data1, jsonp, jsonPrefix) {
      var method;
      this.url = url;
      this.type = type != null ? type : 'GET';
      this.data = data1 != null ? data1 : null;
      this.jsonp = jsonp != null ? jsonp : false;
      this.jsonPrefix = jsonPrefix != null ? jsonPrefix : null;
      this.response = new Response;
      Xhr.COUNTER++;
      if (this.jsonp !== false) {
        if (this.jsonp === true) {
          this.jsonp = 'callback';
        }
        method = Xhr.JSONP_METHOD_PREFIX + Xhr.COUNTER;
        this.url += this.url.indexOf('?') !== -1 ? '&' : '?';
        this.url += this.jsonp + '=' + method;
        window[method] = (function(_this) {
          return function(data) {
            return _this.response.data = data;
          };
        })(this);
      }
      if (this.data !== null) {
        this.data = Helpers.buildQuery(this.data);
        if (this.type !== 'POST') {
          this.url += this.url.indexOf('?') !== -1 ? '&' : '?';
          this.url += this.data;
        }
      }
      this.xhr = this.createXhr();
      this.xhr.open(this.type, this.url, true);
      if (this.url.match(/^(http)s?\:\/\//) === null) {
        this.xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
      }
      if (this.type === 'POST') {
        this.xhr.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
      }
      this.xhr.onreadystatechange = (function(_this) {
        return function() {
          var contentType, data, error, isSuccess, prefix;
          _this.response.state = _this.xhr.readyState;
          if (_this.response.state === 4) {
            _this.response.status = _this.xhr.status;
            isSuccess = (_this.response.status >= 200 && _this.response.status < 300) || _this.response.status === 304;
            if (isSuccess) {
              if (_this.response.status === 204 || _this.type === 'HEAD') {
                _this.response.statusText = 'nocontent';
              } else if (_this.response.status === 304) {
                _this.response.statusText = 'notmodified';
              } else {
                _this.response.statusText = _this.xhr.statusText;
                _this.response.rawData = _this.xhr.responseText;
                _this.response.xml = _this.xhr.responseXML;
                _this.response.data = _this.xhr.responseText;
                contentType = _this.xhr.getResponseHeader('content-type');
                if (contentType !== null && (contentType.match(/application\/json/) !== null || _this.jsonPrefix !== null)) {
                  data = _this.response.data;
                  if (_this.jsonPrefix !== null) {
                    prefix = escape(_this.jsonPrefix);
                    data = data.replace(new RegExp('^' + prefix), '');
                  }
                  _this.response.data = JSON.parse(data);
                }
                if (contentType !== null && (contentType.match(/text\/javascript/) !== null || contentType.match(/application\/javascript/) !== null) && _this.jsonp) {
                  eval(_this.response.data);
                }
              }
              return _this.emit('success', _this.response);
            } else {
              _this.response.statusText = _this.xhr.statusText;
              error = new Error("Can not load " + _this.url + " address");
              error.response = _this.response;
              return _this.emit('error', error, _this.response);
            }
          }
        };
      })(this);
    }

    Xhr.prototype.createXhr = function() {
      if (window.XMLHttpRequest) {
        return new window.XMLHttpRequest;
      } else {
        return new ActiveXObject("Microsoft.XMLHTTP");
      }
    };

    Xhr.prototype.getHeaders = function() {
      return this.xhr.getAllResponseHeaders();
    };

    Xhr.prototype.getHeader = function(name) {
      return this.xhr.getResponseHeader(name);
    };

    Xhr.prototype.setHeader = function(name, value) {
      this.xhr.setRequestHeader(name, value);
      return this;
    };

    Xhr.prototype.setMimeType = function(mime) {
      this.xhr.overrideMimeType(mime);
      return this;
    };

    Xhr.prototype.send = function(fn) {
      this.emit('send', this.response);
      this.on('success', (function(_this) {
        return function(response) {
          _this.emit('complete', null, response);
          return fn(response, null);
        };
      })(this));
      this.on('error', (function(_this) {
        return function(err, response) {
          _this.emit('complete', err, response);
          return fn(null, err);
        };
      })(this));
      this.xhr.send(this.data);
      this.emit('afterSend', this.response);
      return new FakePromise;
    };

    Xhr.prototype.abort = function() {
      this.xhr.abort();
      this.emit('abort', this.response);
      return this;
    };

    return Xhr;

  })(EventEmitter);

  module.exports = Xhr;

}).call(this);
